File: catSurv/R/prior.R
#' Computerized Adaptive Testing Survey Prior Value Function
#'
#' This function returns the prior value for each repondent's position on the latent scale of interest. 
#'
#' @param cat an object of class \code{CATsurv}.
#' @param values vector consisting of quantile for each respondent. 
#' @param name character string for type of assumed distribution. \code{normal} for normal distribution, \code{cauchy} for Cauchy distribtuion, and \code{t} for t-distribution.
#' @param params vector of parameters for selected distribution. The first element for mean value, the second element for standard deviation, and the third element for degree of freedom.
#'
#' @return A vector consisting of prior value for each respondent according to the assumed distribution.
#'  
#' @author Josh W. Cutler and Jacob M. Montgomery
#' @seealso \code{\link{three.pl}},\code{\link{likelihood}}, \code{\link{estimateTheta}}, \code{\link{estimateSE}}, \code{\link{expectedPV}}, \code{\link{nextItem}}, \code{\link{storeAnswer}}, \code{\link{debugNextItem}}
#' @rdname prior.value
#' @export
setGeneric("prior", function(cat, values, name, params){standardGeneric("prior")})

#' @export
setMethod(f="prior", signature=class.name, definition=function(cat, values, name, params) {
  prior.value = switch(name,
                       normal = dnorm(values, params[1], params[2]),
                       cauchy = dcauchy(values, params[1], params[2]),
                       t = 1 / params[2] * dt((values - params[1]) / params[2], df=params[3])
  )
  return(prior.value)
})

File: catSurv/src/Prior/Prior.cpp
#include "Prior.h"

double Prior::dchi(double x, double k) {
	return pdf(chi_squared(k), x);
}

double Prior::dt(double x, double mu, int df) {
	return (dchi(x, df) + mu) / std::sqrt(dnorm(x, 0, 1) / df);
}

double Prior::prior(double x, std::string name, std::vector<double> params) {
	if (name == "normal") {
		return dnorm(x, params[0], params[1]);
	}
	return dt(x, params[0], int(params[1]));
}